### ISSUE TITLE
Angular Universal Prerender Causing Unwanted 301 Redirect

### ISSUE THREAD

**Issue by ewwwgiddings** [angular/universal#2827]
# 🐞 Bug report

#### What modules are related to this issue?

- [ ] builders
- [X] common
- [X] express-engine

### Is this a regression?

No, this has occurred on every version I've used.

### Description

A website you can see this behavior on is https://tragoa.com/welcome

Two explanations of this issue exist at these two SO links:
https://stackoverflow.com/questions/71131171/angular-universal-prerender-causing-301-redirect
https://stackoverflow.com/questions/73666491/issue-with-angular-pre-rendering-with-universal-scully-for-seo

A website I am working on is using Angular Universal 10.1.1 and is hosted on an IIS Server.

If I navigate to the page from the root like foo.com then the website goes straight to the correct route foo.com/welcome without a 301 redirect. However, if I attempt to load foo.com/welcome directly then it will return that request with a 301 redirect to foo.com/welcome/ and that redirect request is returned with a 200 OK and then the trailing slash is stripped in the url bar. So it goes:

1. Request with no trailing slash (/welcome)
2. A 301 redirect returned to request with location including the trailing slash (/welcome/)
3. 200 OK returned
4. Trailing slash stripped in the URL bar in browser

**The main issue here is the unwanted redirect**

This only occurs when the page is using the prerendered HTML. It does not redirect if there is no prerendered index HTML for the given route.

This has caused issues with Lighthouse and some 3rd party redirect issues.

Does anyone know what could be causing this?

Some information that may help:

The base href of the html is `<base href="/">` NOTE: I change this to not have the / then the routes get doubled up like foo.com/welcome/welcome

My server.ts is this:

        import 'zone.js/dist/zone-node';
        
        import { ngExpressEngine } from '@nguniversal/express-engine';
        import * as express from 'express';
        import { join } from 'path';
        import { existsSync, readFileSync } from 'fs';
        
        import 'localstorage-polyfill';
        const domino = require('domino');
        let distFolder = join(process.cwd(), '../App/deploy/spa');
        const template = readFileSync(join(distFolder, 'index.html')).toString();
        const win = domino.createWindow(template);
        win.Object = Object;
        win.Math = Math;
        global['window'] = win;
        global['document'] = win.document;
        global['branch'] = null;
        global['object'] = win.object;
        global['HTMLElement'] = win.HTMLElement;
        global['navigator'] = win.navigator;
        global['localStorage'] = localStorage;
        
        import { AppServerModule } from './src/main.server';
        import { APP_BASE_HREF } from '@angular/common';
        
        export function app(): express.Express {
          const server = express();
          const indexHtml = existsSync(join(distFolder, 'index.original.html')) ? 'index.original.html' : 'index.html';
        
          server.engine('html', ngExpressEngine({
            bootstrap: AppServerModule,
          }));
        
          server.set('view engine', 'html');
          server.set('views', distFolder);
        
          server.get('*.*', express.static(distFolder, {
            maxAge: '1y'
          }));
        
          server.get(
            '*',
            (req, res) => {
            res.render(indexHtml, { req, providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }] });
          });
        
          return server;
        }
        
        function run(): void {
          const port = process.env.PORT || 4000;
        
          const server = app();
          server.listen(port, () => {
            console.log(`Node Express server listening on http://localhost:${port}`);
          });
        }
        
        declare const __non_webpack_require__: NodeRequire;
        const mainModule = __non_webpack_require__.main;
        const moduleFilename = mainModule && mainModule.filename || '';
        if (moduleFilename === __filename || moduleFilename.includes('iisnode')) {
          run();
        }
        
        export * from './src/main.server';

I use this command to create my site and then deploy the files `"prerender": "ng build --configuration production && ng run my-app:prerender:production"`

## 🔬 Minimal Reproduction

To produce the bug I prerender the website using:
`ng build --configuration production && ng run my-app:prerender:production` as the npm prerender command.

Then I put the output directory onto my IIS server and navigate to the root page. This behavior is then observed.

## 🔥 Exception or Error

**The main issue here is the trailing / presence upon redirect**

I do not think that should occur because it affects SEO and creates other issues with loading pages. A big advantage of Universal is the SEO and this minimizes that advantage greatly.

## 🌍 Your Environment

Angular Universal 10.1.1
IIS Server

**Comment by ewwwgiddings** [comment]
I see what's happening is when a request goes to Express without a trailing slash that matches a prerendered route it returns a redirect response to get it at the file path (the same url with a trailing slash). This is evidenced by no redirect when requesting directly with a trailing slash.

For example:
- Navigating to www.website.com/route returns a 301 redirect to location of /route/ if a prerendered file was found at route.
- Navigating directly to www.website.com/route/ leads to no rewrite and returns the prerendered file directly

Is it possible to get around this issue by:

- Altering the requests that do not have a trailing slash in `server.ts`? (i.e. altering requests to have a trailing slash)
- Through anything in Express or Angular's Express Engine? (i.e. strict routing)

There's one workaround at the following link but it does not work for every situation:
https://stackoverflow.com/a/73761714/5832236

**Comment by ewwwgiddings** [comment]
I've tried various things such as:

- Altering the req.url and pathname to include a trailing slash both within a `use()` and the `get()` that includes the render response.
- Using `redirect: false` in the `express.static` options
- Using the `url` parameter in the `render()` method as mentioned here to attempt to add it to the url of the request: https://github.com/angular/universal/tree/main/modules/express-engine

What else can I try? Maybe I'm missing something?

**Comment by ewwwgiddings** [comment]
Here's another SO example. The accepted answer is something I tried but does not work.
https://stackoverflow.com/questions/60210708/angular-universal-prerender-adding-an-extra-slash-on-url

**Comment by ewwwgiddings** [comment]
I'm thinking maybe it's a caching issue? I even changed the redirect option to permanently be false inside of `server-static` and it didn't work.

I started to question if it's maybe because servers aren't picking up on changes inside main.js because the name stays the same? So then I tried to change the server build to use `outputHash: all` but when I did that it gave me this error:
`[error] Error: Could not find the main bundle: C:\Users\ewgid\Desktop\Code\Sites\Tragoa\xTragoaBuildSource1.0.09\TragoaApp\TragoaApp.PUBLISH\deploy\spa\main.js
    at C:\Users\ewgid\Desktop\Code\Sites\Tragoa\xTragoaBuildSource1.0.09\TragoaApp\TragoaApp.SPA\node_modules\@nguniversal\builders\src\prerender\index.js:107:27`

I think it gave me the above error because the prerender only looks for `main.js` not any sort of hashed version of it.

Anyone know how I can bust the cache and get changes inside `main.js` to be recognized if this does sound like a caching issue?
